### PR TITLE
Refactor project packaging and add better testing

### PR DESCRIPTION
**Pull request by WilliamJamieson**
This PR refactors the project to do several things:

1. Use poetry to handle its packaging and dependencies.
2. Add pre-commit to enforce reasonable testing hooks for `git commit`.
3. Remove doc build artifacts.
4. Create basic `tox` environments for testing.
5. Add github actions to enable CI.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+[tool.poetry]
+name = "spacebar"
+version = "0.1.0"
+description = ""
```
Comment (WilliamJamieson) [inline pyproject.toml:4]: @brandon-sexton Please let me know what short description you would like here.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+authors = ["Your Name <you@example.com>"]
```
Comment (WilliamJamieson) [inline pyproject.toml:5]: @brandon-sexton: please let me know what you would like here.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
+name: CI
```
Comment (WilliamJamieson) [inline .github/workflows/ci.yml:1]: @brandon-sexton I think you may need to do some repository setup to get the actions to run (if you want to). I think this set should work.

File: pyproject.toml (hunk 2 above)
Comment (brandon-sexton) [reply pyproject.toml:5]: I'd like something that will properly reflect any collective contributions.  I have no problem making it "Nerd Herd" with a reference to the GitHub team.  What are your thoughts?

**Comment by brandon-sexton** [comment]
I've been busier than normal lately, but I'll look into merging this tonight.  My only concern right now is extending the delay on the next module for the tutorials.

**Comment by WilliamJamieson** [comment]
@brandon-sexton was this merged by accident? I see a revert in PR #4.

**Comment by WilliamJamieson** [comment]
FYI, 99% of the changes in this PR are the fact that the doc build artifacts are being tracked by git. I strongly advise against tracking any automatically generated files/code in git, as this will make diffs really annoying and it tends to cover up all the actual changes (note `black`, which this repository specifically says to run on the code does things to many of those files). Moreover should be unnecessary in general as you can create an action to build/store the doc build somewhere for use by say a github page.

(I haven't looked at the tutorials you mentioned recently) From a teaching perspective, the packaging is important because it should provide a quick start to students being able to actually run and work with the code base. Telling people you need to install this, configure that tends to be user error prone. The packaging should allow users to do a standard `pip install -e .` (`pip install -e ".[test]"` if they want to run/write tests) in their clone and all dependencies should get installed. Moreover, packaging also allows the code to be published to PyPi which makes tutorials built off this code easier to install on the the end user's part.

The rest of the tooling is simply automated development processes, which should be very easy for students to set up (I can in fact write a tutorial for this). As someone who regularly works with code contributed from non-professional software engineers over in `astropy`, getting people used to using lightweight tools that automate things like running `black` and statically analyzing code (`mypy` for example, is used to check the type hints you reference needing in the documentation) is a good idea from the start.

**Comment by WilliamJamieson** [comment]
Note I can break this into smaller chunks if you would like, I purposely did not squash commits to make reviewing this easier.

**Comment by WilliamJamieson** [comment]
I can do a tag-up after 5:00pm EST (today). Just send me an email with the information.

**Pull request merged**